### PR TITLE
Fragment Edit/Add Item

### LINKED ISSUE #12: Fragment Edit/Add Item
Design from Figma

### PR DESCRIPTION
**Pull request by henrijaeger**
closes #12, #6, #4, #3

- [x] Add UI to EditFragment including:
    - [x] name field
    - [x] quantity field
    - [x] dropdown menu for unit
    - [x] best by date incl. DatePicker and localized date format
    - [x] Location Picker incl. Address field, "locate me" / address string parsing functionality and a field for additional information
    - [x] Save button (only displayed in "add mode")

- [x] Save Items in both edit and add "mode"

Things still missing:

- unit tests for UI
- still many hardcoded strings
- unit tests for ItemController, OwnerController and other classes under "datamodel"
add_new_item
edit_item

### CODE REVIEW

**Comment by frcroth** [comment]
@henrijaeger IMO Let's make this ready and merge this as soon as possible, move everything that is still needed into new tickets

File: app/build.gradle [hunk 1]
```diff
@@ -43,6 +43,31 @@ android {
...
+                // We don't have release keys, ignoring
+            }
+            if (keystoreProperties['storeFile']) {
+                storeFile file(keystoreProperties['storeFile'])
+                storePassword keystoreProperties['storePassword']
+                keyAlias keystoreProperties['keyAlias']
+                keyPassword keystoreProperties['keyPassword']
+            }
+        }
+    }
```
Comment (frcroth) [inline app/build.gradle:70]: Is this copied from another source?

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 2]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+        setUpSaveMechanism()
+        setUpOnDateChangedListenerForDatePicker()
+        setUpOnFocusChangeListenerForAddressInputEditText()
+        location = model?.location
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        if (!ADD_ITEM_MODE) {  // **new** Items should not be saved automatically, i. e. onDestroy
+            // TODO: put the following condition into a separate function
```
[the remark this answers is not in the document]
Comment (frcroth) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:115]: With specific hints on what to do, such as this one

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 3]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun setUpOnClickListenersForFormComponents() {
+
+        unitDropdownMenu.setOnMenuItemClickListener { menuItem ->
+            unit_dropdown.editText?.setText(menuItem.title)
+            setModelUnitAttribute()
+            true
+        }
+
+        unit_dropdown.editText?.setOnClickListener() {
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:148]: Is that log still interesting?

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 4]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+
+
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun setUpOnChangedListeners() {
+        itemNameTextInputLayout.editText?.addTextChangedListener { setModelNameAttribute() }
+        itemQuantityTextInputLayout.editText?.addTextChangedListener { setModelQuantityAttribute() }
+        // TODO: uncomment later
+        //itemUnitRadioGroup.setOnCheckedChangeListener { _, _ -> setModelUnitAttribute() }
+        // itemBestByDateTextInputLayout.editText?.addTextChangedListener { setModelBestByDateAttribute() }
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:201]: Remove?

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 5]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+        model?.isShared = itemIsSharedSwitch.isChecked
+    }
+
+
+    private fun setModelDescriptionAttribute() {
+        model?.description = itemDescriptionTextInputLayout.editText?.text.toString()
+    }
+
+    private fun saveNewItem() {
+        // TODO: put the following proofing into a separate function
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:257]: To refactoring issue

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 6]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+                setModelUnitAttribute()
+                setModelBestByDateAttribute()
+                setModelIsSharedAttribute()
+                setModelLocationAttribute()
+                setModelGeohashAttribute()
+                setModelDescriptionAttribute()
+                setModelContactNameAttribute()
+                setModelContactEmailAttribute()
+
+                // TODO: put the following proofing into a separate function
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:280]: Refactoring issue

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 7]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+    private fun buildAddressString(location: GeoPoint): String {
+        val geocoder = Geocoder(requireContext(), Locale.getDefault())
+        val address =
+            geocoder.getFromLocation(location.latitude, location.longitude, 1)
+        val addressLine = address.get(0).getAddressLine(0)
+        val city = address.get(0).locality
+        val state = address.get(0).adminArea
+        val postalCode = address.get(0).postalCode
+
+        return "${addressLine}, ${postalCode} ${city}, ${state}"
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:382]:
```suggestion
        val locality = address.get(0).locality
        val adminArea = address.get(0).adminArea
        val postalCode = address.get(0).postalCode

        return "${addressLine}, ${postalCode} ${locality}, ${adminArea}"
```

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 8]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+    }
+
+    private fun displayAlertOnSaveSharedItemWithoutContactEmail() {
+        AlertDialog.Builder(requireContext())
+            .setTitle("Contact email missing")
+            .setMessage("In order to share an Item, you have to provide a contact email other users can use to reach out to you. Please state your email address in the settings tab.")
+            .setPositiveButton(android.R.string.ok, null)
+            .show()
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:390]: Create issue to remove strings in this file

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 9]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+
+
+    private fun setDateStringToBestByDateEditText() {
+        itemBestByDateTextInputLayout.editText?.setText(buildDateStringFromDatePicker())
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun setDatePickerDateTo(date: Date) {
+        val bestByDate = convertToLocalDate(date)
+        itemBestByDatePicker.updateDate(bestByDate.year, bestByDate.monthValue - 1, bestByDate.dayOfMonth)
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:489]: Why month -1?

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 10]
```diff
@@ -17,39 +51,548 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+        itemAddressTextInputLayout.editText?.isFocusableInTouchMode = false
+        itemAddressTextInputLayout.editText?.inputType = InputType.TYPE_NULL
+        itemAddressTextInputLayout.alpha = .5f
+
+
+        itemDescriptionTextInputLayout.editText?.isEnabled = false
+        itemDescriptionTextInputLayout.editText?.focusable = View.NOT_FOCUSABLE
+        itemDescriptionTextInputLayout.editText?.isFocusableInTouchMode = false
+        itemDescriptionTextInputLayout.editText?.inputType = InputType.TYPE_NULL
+        itemDescriptionTextInputLayout.alpha = .5f
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:562]: Could be refactored -> issue

File: app/src/main/java/app/wefridge/wefridge/NearbyItemFragment.kt [hunk 11]
```diff
@@ -36,15 +38,10 @@ class NearbyItemFragment : Fragment() {
     private var lastVisible: DocumentSnapshot? = null
 
     // TODO: replace with datamodel
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/NearbyItemFragment.kt:40]: TODO done?

File: app/src/main/java/app/wefridge/wefridge/PantryFragment.kt [hunk 12]
```diff
@@ -22,24 +25,43 @@ class PantryFragment : Fragment() {
...
+            displayAlertOnGetItemsFailed()
+        })
+    }
+
+    private fun displayAlertOnGetItemsFailed() {
+        AlertDialog.Builder(requireContext())
+            .setTitle("Error loading your foodstuff")
+            .setMessage("Please check your internet connection and try again.")
+            .setPositiveButton("Retry") { _, _ ->
+                setUpRecyclerViewWithItems()
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/PantryFragment.kt:63]: Strings

File: app/src/main/java/app/wefridge/wefridge/datamodel/Item.kt [hunk 13]
```diff
@@ -0,0 +1,58 @@
+package app.wefridge.wefridge.datamodel
+
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/datamodel/Item.kt:2]: Just a small thing: Why is the directory called "datamodel" and not "model"?

File: app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    /*
+    * The function getItems is based on an example provided on
+    * https://firebase.google.com/docs/firestore/query-data/get-data
+    *
+    * The functions deleteItem, overrideItem and addItem are partially
+    * based on code snippets provided by the Firebase Documentation:
+    * https://firebase.google.com/docs/firestore/manage-data/add-data
+    * */
+    override fun getItems(callbackOnSuccess: (ArrayList<Item>) -> kotlin.Unit, callbackOnFailure: (Exception) -> kotlin.Unit) {
+        // TODO: only get items of specifc user and the groups participants
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt:27]: TODO done?

File: app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        if (item.firebaseId != null) overrideItem(item, { callbackOnSuccess() }, { exception -> callbackOnFailure(exception) })
+        else addItem(item, { callbackOnSuccess() }, { exception -> callbackOnFailure(exception) })
+    }
+
+    private fun overrideItem(item: Item, callbackOnSuccess: () -> kotlin.Unit, callbackOnFailure: (Exception) -> kotlin.Unit) {
+        if (item.firebaseId != null)
+            db.collection("items").document(item.firebaseId!!)
+                .set(item.getHashMap())
+                .addOnSuccessListener {
+                    Log.d(TAG, "item successfully written!")
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt:69]: Maybe add item name in the log, also should not be item successfully written, but "Item successfully overridden"

File: app/src/main/java/app/wefridge/wefridge/placeholder/PlaceholderContent.kt [hunk 16]
```diff
@@ -15,83 +13,10 @@ import java.util.concurrent.ThreadLocalRandom
 object PlaceholderContent {
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/placeholder/PlaceholderContent.kt:14]: Any reason to not delete this file?

Comment (frcroth) [reply app/src/main/java/app/wefridge/wefridge/placeholder/PlaceholderContent.kt:14]: Seems like we don't need placeholders anymore

File: app/build.gradle (hunk 1 above)
Comment (henrijaeger) [reply app/build.gradle:70]: I have copied it from a former branch of @PassiDel. But it's not taken from the internet, so far as I know.

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt (hunk 3 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:149]: No, thank you for noticing that. I have removed it and will commit the change soon.

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt (hunk 9 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:519]: The bestByDate's monthValue attribute ranges from 1 to 12, while the DatePicker expects a range from 0 to 11 to properly display the specified date.

File: app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt (hunk 14 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt:27]: Yes, the first part of this TOD is done.
I would suggest to open a new issue that is about extending the function to deliver items of group participants, too.

File: app/src/main/java/app/wefridge/wefridge/placeholder/PlaceholderContent.kt (hunk 16 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/placeholder/PlaceholderContent.kt:14]: This file is still responsible for storing our items in an ArrayList locally. I thought about moving this Array into the ItemController.
As soon as we know how to handle the ParticipantItems without the PlaceholderContent, we can remove this file.

File: app/src/main/java/app/wefridge/wefridge/datamodel/Item.kt (hunk 13 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/datamodel/Item.kt:2]: As far as I know there's no specific reason. Do you think "model" would be a better name?

Comment (frcroth) [reply app/src/main/java/app/wefridge/wefridge/datamodel/Item.kt:2]: I would prefer it

File: app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt (hunk 14 above)
Comment (frcroth) [reply app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt:27]: Sure

**Review by AlexTerhar — APPROVED** [review]

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt (hunk 8 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:404]: Do you mind if I put this "todo" into the refactoring issue or would you like to have a seperate issue for that?

Comment (frcroth) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:404]: Separate issue was my intention

Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:404]: Okay, I created a new issue for that: #43

File: app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt (hunk 14 above)
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/datamodel/ItemController.kt]: All right, the issue for that is now listed under #44

**Comment by henrijaeger** [comment]
@PassiDel do you also want to review?
@frcroth are you fine with the changes I made since yesterday?
@AlexTerhar thank you for approving!

**Review by frcroth — APPROVED** [review]

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 17]
```diff
@@ -17,39 +52,577 @@ import app.wefridge.wefridge.placeholder.PlaceholderContent
...
+    private var location: GeoPoint? = null
+    private lateinit var sharedPreferences: SharedPreferences
+    private lateinit var unitDropdownMenu: PopupMenu
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+
         arguments?.let {
             model = it.getParcelable(ARG_MODEL)
         }
```
Comment (PassiDel) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:68]: add a blank Item, so it's not nullable and can be used without `model?`

**Review by PassiDel — CHANGES_REQUESTED** [review]

File: app/src/main/res/menu/unit_dropdown.xml [hunk 18]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android">
```
[the remark this answers is not in the document]
Comment (frcroth) [reply app/src/main/res/menu/unit_dropdown.xml:2]: Move to issue

[the remark this answers is not in the document]
Comment (henrijaeger) [reply app/src/main/res/menu/unit_dropdown.xml:2]: The associated issue is now: #46

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 19]
```diff
@@ -1,13 +1,48 @@
 package app.wefridge.wefridge
```
[the remark this answers is not in the document]
Comment (frcroth) [reply app/src/main/java/app/wefridge/wefridge/EditFragment.kt:1]: see #41

**Pull request merged**